### ISSUE TITLE
Device Selection Issue with Ultralytics SAM2: Ignores Specified GPU and Defaults to cuda:0

### ISSUE THREAD

**Issue by jorossi** [ultralytics/ultralytics#21445]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

**Question:**
Version: `8.3.159`  
I'm having trouble with device selection when running a script that uses Ultralytics SAM and PyTorch. I want to use a specific GPU, but the behavior is inconsistent.

Here's what I'm doing:

* I initialize the model like this:

  ```python
  self.model = ultralytics.SAM(model=model_path).to(self.device)
  ```

* Then I run inference like this:

  ```python
  results = self.model(image_path, points=[[x, y]], device=self.device)
  ```

I set `self.device` to `cuda:3`. However, when I run the code with SAM2, the processing splits across two GPUs. Some data loads into GPU 3, but most of the work happens on GPU 0. Here’s the output from Torch debug:

```
[INFO] SamSegmenter.__init__: Using device: cuda:0
[INFO] SamSegmenter.__init__: CUDA available: True
[INFO] SamSegmenter.__init__: CUDA device count: 8
[INFO] SamSegmenter.__init__: Current CUDA device: 0
```
Why is SAM defaulting to `cuda:0` even though I explicitly set `cuda:3` during both model loading and inference?

When I use YOLO with the same setup, the device assignment works as expected. The only workaround I found is setting an environment variable, but I'd prefer not to do that. Is there a better way?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @jorossi, thank you for bringing this to our attention and for your detailed description! 🚀 This is an automated response to help guide you efficiently—an Ultralytics engineer will review your issue and provide more specific assistance soon.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for further details on device selection and usage, as well as checking the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples for best practices.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can investigate your SAM2 device selection issue more effectively.

For custom training or inference questions, please include all relevant configuration details, sample code, and logs, and verify that you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community to connect with others:
- For real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussion: [Discourse](https://community.ultralytics.com/)
- To share knowledge: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you are using the latest `ultralytics` package (with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to confirm your issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience and for helping us improve Ultralytics!

**Comment by Y-T-G** [comment]
You can try forcing your device with `torch.cuda.device()` context manager:

```python
with torch.cuda.device(3):
    results = model.predict(device=torch.device(3))
```

**Comment by jorossi** [comment]
Hi, thanks for the quick reply! The issue is resolved now using the fix you suggested.
I couldn't figure it out myself, so I really appreciate your help, this finally works.

For reference, I didn't need to specify the device when initializing the model. Changing the call to this solved it:

```python
self.model = ultralytics.SAM(model=model_path)

with torch.cuda.device(self.device):
    results = self.model(
        image_path,
        points=[[x, y]],
        device=torch.device(f'cuda:{self.device}'),
    )
```
It now uses the specified GPU index as expected.

**Comment by glenn-jocher** [comment]
Great to hear the issue is resolved! The `torch.cuda.device()` context manager was the right approach here because SAM models have more complex internal device handling compared to standard YOLO models, and this ensures all operations within that context use the specified device consistently.

Thanks for sharing your working solution - it will be helpful for other users encountering similar device selection issues with SAM models.
